modules/core/04-channel/keeper/upgrade.go

```diff
@@ -0,0 +1,91 @@
+package keeper
+
+import (
+	"reflect"
+
+	errorsmod "cosmossdk.io/errors"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	connectiontypes "github.com/cosmos/ibc-go/v7/modules/core/03-connection/types"
+	"github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v7/modules/core/24-host"
+)
+
+// ChanUpgradeInit is called by a module to initiate a channel upgrade handshake with
+// a module on another chain.
+func (k Keeper) ChanUpgradeInit(
+	ctx sdk.Context,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	proposedUpgradeChannel types.Channel,
+	counterpartyTimeoutHeight clienttypes.Height,
+	counterpartyTimeoutTimestamp uint64,
+) (uint64, string, error) {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return 0, "", errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if channel.State != types.OPEN {
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.OPEN, channel.State)
+	}
+
+	if !k.scopedKeeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
+		return 0, "", errorsmod.Wrapf(types.ErrChannelCapabilityNotFound, "caller does not own capability for channel, port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	channel.State = types.INITUPGRADE
+	if reflect.DeepEqual(channel, proposedUpgradeChannel) {
+		return 0, "", errorsmod.Wrap(types.ErrChannelExists, "existing channel end is identical to proposed upgrade channel end")
+	}
+
+	if !k.connectionKeeper.HasConnection(ctx, proposedUpgradeChannel.ConnectionHops[0]) {
+		return 0, "", errorsmod.Wrapf(connectiontypes.ErrConnectionNotFound, "failed to retrieve connection: %s", proposedUpgradeChannel.ConnectionHops[0])
+	}
+
+	if proposedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
+		proposedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidCounterparty, "counterparty port ID and channel ID cannot be upgraded")
+	}
+
+	if !channel.Ordering.SubsetOf(proposedUpgradeChannel.Ordering) {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
+	}
+
+	upgradeSequence := uint64(1)
+	if seq, found := k.GetUpgradeSequence(ctx, portID, channelID); found {
+		upgradeSequence = seq + 1
+	}
+
+	upgradeTimeout := types.UpgradeTimeout{
+		TimeoutHeight:    counterpartyTimeoutHeight,
+		TimeoutTimestamp: counterpartyTimeoutTimestamp,
+	}
+
+	k.SetUpgradeRestoreChannel(ctx, portID, channelID, channel)
```

Comment:
channel will have a state of `types.INITUPGRADE` (above on line 41), the channel stored here should be the channel before any modifications were made right?

Reply:
I've changed it to the following: 

```go
restoreChannel := channel
channel.State = types.INITUPGRADE
if reflect.DeepEqual(channel, proposedUpgradeChannel) {
	return 0, "", errorsmod.Wrap(types.ErrChannelExists, "existing channel end is identical to proposed upgrade channel end")
}
```

I don't really like changing the state of the channel just to do a deep equals comparison. Easy to lead to bugs like what you just caught. But we can worry about code vanity later on this one I guess.

Reply:
LGTM thanks for the quick fix